packages/kbn-esql-ast/src/visitor/contexts.ts

```diff
@@ -0,0 +1,437 @@
+
+    /**
+     * ES|QL AST node which is currently being visited.
+     */
+    public readonly node: Node,
+
+    /**
+     * Context of the parent node, from which the current node was reached
+     * during the AST traversal.
+     */
+    public readonly parent: VisitorContext | null = null
+  ) {}
+
+  public *visitArguments(
+    input: ExpressionVisitorInput<Methods>
+  ): Iterable<ExpressionVisitorOutput<Methods>> {
+    this.ctx.assertMethodExists('visitExpression');
+
+    const node = this.node;
+
+    if (!isNodeWithArgs(node)) {
+      throw new Error('Node does not have arguments');
+    }
+
+    for (const arg of singleItems(node.args)) {
+      yield this.visitExpression(arg, input as any);
+    }
+  }
+
+  public visitExpression(
+    expressionNode: ESQLAstExpressionNode,
+    input: ExpressionVisitorInput<Methods>
+  ): ExpressionVisitorOutput<Methods> {
+    return this.ctx.visitExpression(this, expressionNode, input);
+  }
+
+  public visitCommand(
+    commandNode: ESQLAstCommand,
+    input: ExpressionVisitorInput<Methods>
+  ): ExpressionVisitorOutput<Methods> {
+    return this.ctx.visitCommand(this, commandNode, input);
+  }
+}
+
+export class QueryVisitorContext<
+  Methods extends VisitorMethods = VisitorMethods,
+  Data extends SharedData = SharedData
+> extends VisitorContext<Methods, Data, ESQLAstQueryNode> {
+  public *visitCommands(
+    input: UndefinedToVoid<Parameters<NonNullable<Methods['visitCommand']>>[1]>
+  ): Iterable<
+    | ReturnType<NonNullable<Methods['visitCommand']>>
+    | ReturnType<NonNullable<Methods['visitFromCommand']>>
+  > {
+    this.ctx.assertMethodExists('visitCommand');
+
+    for (const cmd of this.node) {
+      yield this.visitCommand(cmd, input as any);
+    }
+  }
+}
+
+// Commands --------------------------------------------------------------------
+
+export class CommandVisitorContext<
+  Methods extends VisitorMethods = VisitorMethods,
+  Data extends SharedData = SharedData,
+  Node extends ESQLAstCommand = ESQLAstCommand
+> extends VisitorContext<Methods, Data, Node> {
+  public name(): string {
+    return this.node.name.toUpperCase();
+  }
+
+  public *options(): Iterable<ESQLCommandOption> {
+    for (const arg of this.node.args) {
+      if (Array.isArray(arg)) {
+        continue;
+      }
+      if (arg.type === 'option') {
+        yield arg;
+      }
+    }
+  }
+
+  public *visitOptions(
+    input: UndefinedToVoid<Parameters<NonNullable<Methods['visitCommandOption']>>[1]>
+  ): Iterable<ReturnType<NonNullable<Methods['visitCommandOption']>>> {
+    this.ctx.assertMethodExists('visitCommandOption');
+
+    for (const option of this.options()) {
+      const sourceContext = new CommandOptionVisitorContext(this.ctx, option, this);
+      const result = this.ctx.methods.visitCommandOption!(sourceContext, input);
+
+      yield result;
+    }
+  }
+
+  public *arguments(option: '' | string = ''): Iterable<ESQLAstItem> {
```

Comment:
Slick to use generators here

Reply:
Yeah, generators/iterators allow to walk only as much as you need and then break out of the loop with `break;`.

Reply:
yes very cool to see this!